Config 2.0.0

ntsanov opened the pull request:
## Issue

https://github.com/harmony-one/bounties/issues/34

## Test

### Unit Test Coverage

Before:

```
PASS
coverage: 46.0% of statements
ok      github.com/harmony-one/harmony/cmd/harmony      0.043s
```

After:

```
PASS
coverage: 47.1% of statements
ok      github.com/harmony-one/harmony/cmd/harmony      0.072s
```

### Test/Run Logs
When running and stdin is terminal
```
Loaded config version 1.0.4 which is not latest (2.0.0).
Do you want to update config to latest version: (Y/n)

Config updated from 1.0.4 to 2.0.0
```
When stdin is piped
```
echo Y |./harmony -c config.toml
Loaded config version 1.0.4 which is not latest (2.0.0).
Update saved config with `./harmony config update [config_file]`
```

## Operational Checklist

1. **Does this PR introduce backward-incompatible changes to the on-disk data structure and/or the over-the-wire protocol?**. (If no, skip to question 8.)

    **NO**

8. **Does this PR introduce backward-incompatible changes *NOT* related to on-disk data structure and/or over-the-wire protocol?** (If no, continue to question 11.)

    **NO**

JackyWYX commented:
Is this PR ready for review or still WIP?

ntsanov commented:
> Is this PR ready for review or still WIP?

It is ready for review. 

Just to point out - one key factor that might be worth discussing  is what to be done with file versions that are not in the migrations map. Currently all versions that are not found are being handled with same migration as of 1.0.4. Some of the other options would be:
 - adding migrations for previous versions since there probably are not that many previous versions
 - comparing the with current and older versions are migrated from the first migration available while for newer versions (probably old version bin with new config) either error is returned or warning and loading the default

Another is whether flags should match exactly config like dnssync.port or should stay as they are dns.port . Currently they are as they were ex - dns.port

JackyWYX commented on cmd/harmony/config_migrations.go:
```diff
@@ -0,0 +1,81 @@
+	migration, found := migrations[confVersion]
+	// Legacy config versions
+	if !found {
+		migration = migrations["1.0.4"]
```
Probably show some warning message in this case?

ntsanov replied:
Will add a warning message

JackyWYX commented on cmd/harmony/flags.go:
```diff
@@ -392,20 +392,20 @@ func applyNetworkFlags(cmd *cobra.Command, cfg *harmonyConfig) {
 
 	if cli.IsFlagChanged(cmd, dnsZoneFlag) {
-		cfg.Network.DNSZone = cli.GetStringFlagValue(cmd, dnsZoneFlag)
+		cfg.DNSSync.Zone = cli.GetStringFlagValue(cmd, dnsZoneFlag)
```
Since the field has already been changed, do you think it would be better to aggragate the dns related flags to the same catagory?

JackyWYX commented on cmd/harmony/main.go:
```diff
@@ -159,23 +161,43 @@ func raiseFdLimits() error {
+		if stat.Mode()&os.ModeCharDevice != 0 {
+			var readStr string
+			fmt.Println("Do you want to update config to latest version: (Y/n)")
+			fmt.Scanf("%s", &readStr)
```
Can we add a timeout to the `Scanf` to better support `node.sh`? The shell script was recommended to run in tmux session long time ago, and we need to take std input timeout as the corner case for backward compatibility.

Also, I would appreciate if you can refactor this piece of code to a different function.

ntsanov replied:
Timeout seems the perfect way to make absolutely sure It won't get stuck there ... should have thought of that

JackyWYX commented on cmd/harmony/main.go:
```diff
@@ -159,23 +161,43 @@ func raiseFdLimits() error {
+			}
+			readStr = strings.ToLower(readStr)
+			if readStr == "y" || readStr == "" {
+				writeHarmonyConfigToFile(config, configFile)
```
Would it be better to leave a copy of the older config file being replaced?

ntsanov replied:
Yes, same like config update, will move it to <config_path>.backup

JackyWYX commented:
Thanks for your effort in this bounty. 

#### For your questions:

> one key factor that might be worth discussing is what to be done with file versions that are not in the migrations map. Currently all versions that are not found are being handled with same migration as of 1.0.4. Some of the other options would be:
> * adding migrations for previous versions since there probably are not that many previous versions
> * comparing the with current and older versions are migrated from the first migration available while for newer versions (probably old version bin with new config) either error is returned or warning and loading the default.

Adding migrations between v1.0.0 to v1.0.4 is optional - they can be treated as v1.0.4 for all the hackings to make them backward compatible. But it would be definitely better to include previous version of config in the migration unit tests.

> Another is whether flags should match exactly config like dnssync.port or should stay as they are dns.port . Currently they are as they were ex - dns.port

`--dns.port` is fine to be left as it is. But some other flags would be better to be renames, E.g. `sync.server` -> `dns.server`, e.t.c. Please also note that in the PR just merged (https://github.com/harmony-one/harmony/pull/3669), `--dns.port` was changed default value from 9000 to 9000-3000=6000. Please be notified of the change

For newer versions, it would be definitely better to warn user about the version issue, and then load the config as latest in binary. So would you please introduce [go-version](https://github.com/hashicorp/go-version) for version string validation and comparison? 

#### For PR reviews: generally looks good to me. Some comments:

1. Please fix the build errors to have this PR passed in travis CICD test.
2. Please rebase this branch on main after merging https://github.com/harmony-one/harmony/pull/3669. In this PR, v1.0.5 is introduced to update port number in `network.DNSPort` (9000 -> 6000), and added `sync.legacyServer.Port` (defValue: 6000). I am sorry for not including this PR in the bounty description in the first place. BTW, since v1.0.5 is not released, so you are free to integrate it along with v2.0.0 update.
3. For flags, it's recommended to update the fields if it makes sense. For example, `--dns.port` -> `--dns.zone.port`, `--sync.legacy.server` -> `--dns.server`, e.t.c. It is also recommended to have deprecated flags for maximum backward compatibility.
4. Do you mind dump the previous version toml files to integrate them with unit tests?

ntsanov commented on cmd/harmony/config.go:
```diff
@@ -35,15 +35,26 @@ type harmonyConfig struct {
+	Port             int    // replaces: Network.DNSSyncPort
+	Zone             string // replaces: Network.DNSZone
+	LegacySyncing    bool   // replaces: Network.LegacySyncing
+	LegacyClient     bool   // replaces: Sync.LegacyClient
```
Understood. I removed Legacy from everything initially but reading some logic about LegacySyncing got me confused and I returned them

ntsanov commented on cmd/harmony/config.go:
```diff
@@ -35,15 +35,26 @@ type harmonyConfig struct {
-	LegacySyncing bool // if true, use LegacySyncingPeerProvider
-	DNSZone       string
-	DNSPort       int
+	// TODO remove
```
Of course, I guess I missed this one - will check the code again

ntsanov commented on cmd/harmony/config.go:
```diff
@@ -268,10 +299,40 @@ func getDefaultSyncConfig(nt nodeconfig.NetworkType) syncConfig {
+			fmt.Println(err)
+			os.Exit(128)
+		}
+		config, migratedVer, err := migrateConf(configBytes)
```
Yes, seems the right thing to do, I would move it to <config_path>.backup

ntsanov commented on cmd/harmony/config_migrations.go:
```diff
@@ -0,0 +1,81 @@
+type configMigrationFunc func(*toml.Tree) *toml.Tree
+
+func init() {
+	migrations["1.0.4"] = func(conf *toml.Tree) *toml.Tree {
```
> Just curious: Is go-toml package able to set a certain field to nil?

Well, you can delete a key I guess, but that would take whatever is the default value of that field when it us unmarshaled

> Also is there a more systematic approach if we can upgrade a config with older version maximizing reusage of the code? E.g, upgrade config from "v1.0.4" -> "v2.0.0" -> "v3.0.0"?

This is exactly what I had in mind. The idea is that each migration contains **only** the necessary changes that will migrate it to the next version. Migrations are then applied until the config version matches tomlConfigVersion const. In the example of "v1.0.4" -> "v2.0.0" -> "v3.0.0" we already have  "1.0.4"  that migrates config to 2.0.0 so we only need to define migration  "2.0.0" that does all the transformations to 3.0.0

ntsanov commented on cmd/harmony/config_migrations.go:
```diff
@@ -0,0 +1,81 @@
+	if confVersion == "" {
+		return harmonyConfig{}, "", errors.New("invalid config file - version missing")
+	}
+	migratedVer := confVersion
```
> Thanks for your effort in this bounty.
> 
> #### For your questions:
> > one key factor that might be worth discussing is what to be done with file versions that are not in the migrations map. Currently all versions that are not found are being handled with same migration as of 1.0.4. Some of the other options would be:
> > 
> > * adding migrations for previous versions since there probably are not that many previous versions
> > * comparing the with current and older versions are migrated from the first migration available while for newer versions (probably old version bin with new config) either error is returned or warning and loading the default.
> 
> Adding migrations between v1.0.0 to v1.0.4 is optional - they can be treated as v1.0.4 for all the hackings to make them backward compatible. But it would be definitely better to include previous version of config in the migration unit tests.

Ok, so I will build some old versions and dump the config and add it to the test cases

> 
> > Another is whether flags should match exactly config like dnssync.port or should stay as they are dns.port . Currently they are as they were ex - dns.port
> 
> `--dns.port` is fine to be left as it is. But some other flags would be better to be renames, E.g. `sync.server` -> `dns.server`, e.t.c. Please also note that in the PR just merged (#3669), `--dns.port` was changed default value from 9000 to 9000-3000=6000. Please be notified of the change
> 
> For newer versions, it would be definitely better to warn user about the version issue, and then load the config as latest in binary. So would you please introduce [go-version](https://github.com/hashicorp/go-version) for version string validation and comparison?

Ok, will use go-version to make sure that it is a valid semantic version. Older then 1.0.4 will be handled as 1.0.4 while any unknown version will print warning and load the defaults

> 
> #### For PR reviews: generally looks good to me. Some comments:
> 1. Please fix the build errors to have this PR passed in travis CICD test.
> 2. Please rebase this branch on main after merging #3669. In this PR, v1.0.5 is introduced to update port number in `network.DNSPort` (9000 -> 6000), and added `sync.legacyServer.Port` (defValue: 6000). I am sorry for not including this PR in the bounty description in the first place. BTW, since v1.0.5 is not released, so you are free to integrate it along with v2.0.0 update.

Just to clarify this - any config prior to 2.0.0 containing network.DNSPort == 9000 should be changed to 6000 ? If so that can also be part of the migration

> 3. For flags, it's recommended to update the fields if it makes sense. For example, `--dns.port` -> `--dns.zone.port`, `--sync.legacy.server` -> `--dns.server`, e.t.c. It is also recommended to have deprecated flags for maximum backward compatibility.
> 4. Do you mind dump the previous version toml files to integrate them with unit tests?

Ok I will rebase and resolve the issues that has been discussed so far

ntsanov commented:
> Thanks for your effort in this bounty.
> 
> #### For your questions:
> > one key factor that might be worth discussing is what to be done with file versions that are not in the migrations map. Currently all versions that are not found are being handled with same migration as of 1.0.4. Some of the other options would be:
> > 
> > * adding migrations for previous versions since there probably are not that many previous versions
> > * comparing the with current and older versions are migrated from the first migration available while for newer versions (probably old version bin with new config) either error is returned or warning and loading the default.
> 
> Adding migrations between v1.0.0 to v1.0.4 is optional - they can be treated as v1.0.4 for all the hackings to make them backward compatible. But it would be definitely better to include previous version of config in the migration unit tests.

Ok, so I will build some old versions and dump the config and add it to the test cases

> 
> > Another is whether flags should match exactly config like dnssync.port or should stay as they are dns.port . Currently they are as they were ex - dns.port
> 
> `--dns.port` is fine to be left as it is. But some other flags would be better to be renames, E.g. `sync.server` -> `dns.server`, e.t.c. Please also note that in the PR just merged (#3669), `--dns.port` was changed default value from 9000 to 9000-3000=6000. Please be notified of the change
> 
> For newer versions, it would be definitely better to warn user about the version issue, and then load the config as latest in binary. So would you please introduce [go-version](https://github.com/hashicorp/go-version) for version string validation and comparison?

Ok, will use go-version to make sure that it is a valid semantic version. Older then 1.0.4 will be handled as 1.0.4 while any unknown version will print warning and load the defaults

> 
> #### For PR reviews: generally looks good to me. Some comments:
> 1. Please fix the build errors to have this PR passed in travis CICD test.
> 2. Please rebase this branch on main after merging #3669. In this PR, v1.0.5 is introduced to update port number in `network.DNSPort` (9000 -> 6000), and added `sync.legacyServer.Port` (defValue: 6000). I am sorry for not including this PR in the bounty description in the first place. BTW, since v1.0.5 is not released, so you are free to integrate it along with v2.0.0 update.

Just to clarify this - any config prior to 2.0.0 containing network.DNSPort == 9000 should be changed to 6000 ? If so that can also be part of the migration

> 3. For flags, it's recommended to update the fields if it makes sense. For example, `--dns.port` -> `--dns.zone.port`, `--sync.legacy.server` -> `--dns.server`, e.t.c. It is also recommended to have deprecated flags for maximum backward compatibility.
> 4. Do you mind dump the previous version toml files to integrate them with unit tests?

Ok I will rebase and resolve the issues that has been discussed so far

ntsanov commented:
This is a list of the thing I have done after what we discussed
 - Added timeout to config update prompt. Prompt logic has been moved to
 a separate func
 - When updating config file, the original is saved to <config>.backup
 - Added semantic version validation for the version found in config. Error is
 returned if version is missing, not a valid sematic version or not
 found in the migrations map
 - Flags related to DNSSync has been moved to dnsSyncFlags flag array.
 Also moved tests related to DNSSync flags to separate test func
 - Added dns.server and dns.client flags and tests for them
 - Added migration tests with dumps of versions 1.0.2, 1.0.3, 1.0.4

I've also moved the logic for legayDNSPort from correctLegacyHarmonyConfig to the migration so I can get rid of that legayDNSPort field in Network, but have left the other logic in correctLegacyHarmonyConfig untouched.

I've  done the rebase, hope I havn't missed something  since there were some relevant changes

Still waiting for CICD to finish

Thanks for your time reviewing the code

JackyWYX reviewed (commented):
Thanks for your effort! It's really amazing work. I especially like the following change which makes the nasty v1.0.5 update looks neat:

```
I've also moved the logic for legayDNSPort from correctLegacyHarmonyConfig to the migration so I can get rid of that legayDNSPort field in Network, but have left the other logic in correctLegacyHarmonyConfig untouched.
```

There are only some comments on minor issues. 

Other than that, I would suggest to give a system test report on what config and flags are used and what is observed (You need to have block number increase as the test pass)

Excellent work! Thanks for your time.

ntsanov commented on cmd/harmony/config.go:
```diff
@@ -288,28 +357,62 @@ func registerDumpConfigFlags() error {
+			readStr = readStr[0:1]
+		}
+		readStr = strings.ToLower(readStr)
+		return readStr == "y" || readStr == ""
```
My thoughts were that the user was given a choice [Y/n] with yes being the default , so pressing enter without entering anything would be considered as taking the default. Shall I change it to default to no [y/N] ?

ntsanov replied:
Well I did change it to default to N, if need be,  can be switched back

ntsanov commented on cmd/harmony/config_migrations.go:
```diff
@@ -0,0 +1,116 @@
+		var port = int64(nodeconfig.DefaultDNSPort)
+		portField := confTree.Get("Network.DNSPort")
+		if p, ok := portField.(int64); ok {
+			if p != nodeconfig.DefaultLegacyDNSPort {
```
Yes, you are right, will fix it and add a test for this case. Just to be absolutely clear - we subtract 3000 from legacy configs with the exception of the port being == DefaultDNSPort which is 6000 at the moment ?

JackyWYX replied:
It is not expected that any node will run legacy port as 6000. Even if they config the port as 6000, it will be treated as 3000. So it's just fine to subtract 3000 for all cases.

ntsanov replied:
> * For each test case, have config file and cli command containing flags as test data.
> * Running CLI command with the config file will give expected node running behavior. (E.g., run with dns.serverPort will have node listen on specified server port. You can simply print that message in a customized branch.)
> * All test cases shall be reproducible.
> 
> I did some test cases long time ago, probably you can take a look:
> 
> * https://gist.github.com/JackyWYX/2554d561ccca74617d68a74a53ca64e3
> * https://gist.github.com/JackyWYX/adae08dd1984a2943c10a5358aa4b59a
> 

Thank you, that was helpful. I think I have a general Idea now what is required. I will conduct the tests today and post the result

ntsanov commented on api/service/legacysync/downloader/server.go:
```diff
@@ -62,7 +64,7 @@ func (s *Server) Start(ip, port string) (*grpc.Server, error) {
 // NewServer creates new Server which implements DownloadInterface.
-func NewServer(dlInterface DownloadInterface) *Server {
-	s := &Server{downloadInterface: dlInterface}
+func NewServer(dlInterface DownloadInterface, port int) *Server {
```
I did the rebase again

ntsanov commented on cmd/harmony/config.go:
```diff
@@ -167,6 +172,29 @@ type syncConfig struct {
 	DiscBatch      int  // size of each discovery
 }
 
+// sanityFixHarmonyConfig correct values for old config version load
```
Done

ntsanov commented on cmd/harmony/config.go:
```diff
@@ -288,28 +357,62 @@ func registerDumpConfigFlags() error {
-	// Correct for old config version load (port 0 is invalid anyways)
-	if config.HTTP.RosettaPort == 0 {
-		config.HTTP.RosettaPort = defaultConfig.HTTP.RosettaPort
+	correctLegacyHarmonyConfig(&config)
```
Done

JackyWYX replied:
Would you please move this function to be only used when loading version <=v1.0.4? These are nasty "hack" solutions, and we don't want this value correction for future updates. Really appreciated.

ntsanov replied:
I have left correctLegacyHarmonyConfig intact since I wasn't sure it is not used to cover other cases (like user deleting part of the config), but In that case I  moved all the logic in 1.0.4 migration and remove correctLegacyHarmonyConfig altogether

JackyWYX replied:
I gotta. Actually the content in `correctLegacyHarmonyConfig` is only used for v1.0.4. And `correctHarmonyConfig` shall be applied to all versions, which we shall add this function later.

You also provided a very good insight: what if a user deleted part of the config which is essential for starting the node? Shall we give it a default value and warn user of the absence of config data? Maybe I can create another bounty for this one later :)

ntsanov replied:
Yes it does have some use cases. User would be able to used very short config files. Basically your config file could only contain a few parameters that you want to override, everything else uses default values

ntsanov commented on cmd/harmony/config_migrations.go:
```diff
@@ -0,0 +1,116 @@
+		return harmonyConfig{}, "", errors.New("config file invalid - no version entry found")
+	}
+	migratedFrom = confVersion
+	if confVersion != tomlConfigVersion {
```
Moved the logic to doMigrations func

ntsanov commented on cmd/harmony/default.go:
```diff
@@ -130,9 +130,9 @@ var (
-		LegacyServer:   true,
-		LegacyClient:   true,
+		Downloader: false,
+		// LegacyServer:   true, // moved to dnsSync
```
Done

ntsanov commented on cmd/harmony/default.go:
```diff
@@ -143,9 +143,9 @@ var (
-		LegacyServer:   true,
-		LegacyClient:   false,
+		Downloader: true,
+		// LegacyServer:   true, // moved to dnsSync
```
Done

ntsanov commented on cmd/harmony/config_migrations.go:
```diff
@@ -0,0 +1,116 @@
+		serverPortField := confTree.Get("Sync.LegacyServerPort")
+		if serverPort, ok := serverPortField.(int64); ok {
+			confTree.Set("DNSSync.ServerPort", serverPort)
+		}
```
Yes that makes sense. I see now why I didn't find it in the initial version. Have added check if the key is missing to set it to nodeconfig.DefaultDNSPort

ntsanov commented:
I've done the changes we discussed, should all be resolved now. I've also done the rebase. Is it ok now ?

> Other than that, I would suggest to give a system test report on what config and flags are used and what is observed (You need to have block number increase as the test pass)

Is there a guideline or template for that? Should I just run a node with default settings and report the logs and config/flag options I ran it with?

> Excellent work! Thanks for your time.

Thank you also, I am really impressed by the fast reaction and accurate assessment of the code

JackyWYX commented on cmd/harmony/config_migrations.go:
```diff
@@ -0,0 +1,131 @@
+		if port, ok := serverPortField.(int64); ok {
+			serverPort = int(port)
+		}
+		confTree.Set("DNSSync.ServerPort", serverPort)
```
Probably in else bracket for not overwriting the customized port?

JackyWYX commented:
> I've done the changes we discussed, should all be resolved now. I've also done the rebase. Is it ok now ?

Only one issue noticed: https://github.com/harmony-one/harmony/pull/3681#discussion_r624492213. Please put `correctLegacyHarmonyConfig` function to be only used for v1.0.4. This is ugly and legacy compared with your elegant solution :) 

> Is there a guideline or template for that? Should I just run a node with default settings and report the logs and config/flag options I ran it with?

What's in my mind is just to having some different test cases which is reproducible: 
* For each test case, have config file and cli command containing flags as test data.
* Running CLI command with the config file will give expected node running behavior. (E.g., run with dns.serverPort will have node listen on server port. You can simply print that message in a customized branch.)
* All test cases shall be reproducible.

I did some test cases long time ago, probably you can take a look: https://gist.github.com/JackyWYX/adae08dd1984a2943c10a5358aa4b59a. 

I will take a final pass on the code later. Hopefully we can merge early next week :)

ntsanov commented:
> * For each test case, have config file and cli command containing flags as test data.
> * Running CLI command with the config file will give expected node running behavior. (E.g., run with dns.serverPort will have node listen on specified server port. You can simply print that message in a customized branch.)
> * All test cases shall be reproducible.
> 
> I did some test cases long time ago, probably you can take a look:
> 
> * https://gist.github.com/JackyWYX/2554d561ccca74617d68a74a53ca64e3
> * https://gist.github.com/JackyWYX/adae08dd1984a2943c10a5358aa4b59a
> 

Thank you, that was helpful. I think I have a general Idea now what is required. I will conduct the tests today and post the result

ntsanov commented:
Test result are acquired by building a from  branch - https://github.com/ntsanov/harmony/tree/print-values that has additional Print lines added to node_syncing constructors

## Test case 1 - Starting harmony without any arguments or config
Running:
```./harmony```

Result: 
```Old config version detected 
Do you want to update config to the latest version: [y/N]
N
Staking mode; node key feea6d16f9073c4a7c8c55cd18822cece32c471bb389b472b884d5cec9e207f5086dc44735d5dc45c4069879383fcf14; -> shard 0
DNSSyncing: true
Syncing Port: 3000
Syncing Zone: t.hmny.io
GRPC Sync Server: true
GRPC Sync Server Port:6000 
GRPC Sync Client: true
Started RPC server at: 127.0.0.1:9500
Started WS server at: 127.0.0.1:9800
```
Problem Found:  When no config set, migratedFrom was also not set so all version was considered old and was asking for it to be updated.
Solved: When using default config and not loading from file, set isUsingDefault boolean variable and only ask if false

## Test case 2 - Passing dns.port, dns.zone
Running:

```
./harmony --dns.port=9001 --dns.zone=127.0.0.1 
```

Result:
```
Staking mode; node key feea6d16f9073c4a7c8c55cd18822cece32c471bb389b472b884d5cec9e207f5086dc44735d5dc45c4069879383fcf14; -> shard 0
DNSSyncing: true
Syncing Port: 6001
Syncing Zone: 127.0.0.1
GRPC Sync Server: true
GRPC Sync Server Port:6000 
GRPC Sync Client: true
Started RPC server at: 127.0.0.1:9500
Started WS server at: 127.0.0.1:9800
```
Problems Found: None


## Test case 3 Running with config 1.0.2
Running:

```./harmony -c config_1_0_2.toml```

Result:
```
Old config version detected 1.0.2
Do you want to update config to the latest version: [y/N]

Staking mode; node key feea6d16f9073c4a7c8c55cd18822cece32c471bb389b472b884d5cec9e207f5086dc44735d5dc45c4069879383fcf14; -> shard 0
DNSSyncing: true
Syncing Port: 3000
Syncing Zone: t.hmny.io
GRPC Sync Server: true
GRPC Sync Server Port:6000 
GRPC Sync Client: true
Started RPC server at: 127.0.0.1:9500
Started WS server at: 127.0.0.1:9800
```
Possible Problem: It seems that port DNSSync is actually  port - 6000. Not sure if that is intended but we subtract 3000 in the migration, but looking at the above tests, there seem to be somewhere else in the code were there is also - 3000 subtraction

## Test case 4 Running with config 1.0.3

Running:

```./harmony -c config_1_0_3.toml```

Result:
```
Old config version detected 1.0.3
Do you want to update config to the latest version: [y/N]

Staking mode; node key feea6d16f9073c4a7c8c55cd18822cece32c471bb389b472b884d5cec9e207f5086dc44735d5dc45c4069879383fcf14; -> shard 0
DNSSyncing: true
Syncing Port: 3000
Syncing Zone: t.hmny.io
GRPC Sync Server: true
GRPC Sync Server Port:6000 
GRPC Sync Client: true
Started RPC server at: 127.0.0.1:9500
Started WS server at: 127.0.0.1:9800
```
Problems Found: None (Same applies for test 3 and port-6000)

## Test case 5 Running with config 1.0.4
Running:

```./harmony -c config_1_0_4.toml```

Result:
```
ld config version detected 1.0.4
Do you want to update config to the latest version: [y/N]

Staking mode; node key feea6d16f9073c4a7c8c55cd18822cece32c471bb389b472b884d5cec9e207f5086dc44735d5dc45c4069879383fcf14; -> shard 0
DNSSyncing: true
Syncing Port: 3000
Syncing Zone: t.hmny.io
GRPC Sync Server: true
GRPC Sync Server Port:6000 
GRPC Sync Client: true
Started RPC server at: 127.0.0.1:9500
Started WS server at: 127.0.0.1:9800
```
Problems Found: None (Same applies for test 3 and port-6000)

JackyWYX commented:
> Test result are acquired by building a from branch - https://github.com/ntsanov/harmony/tree/print-values that has additional Print lines added to node_syncing constructors
> 
> ## Test case 1 - Starting harmony without any arguments or config
> Running:
> `./harmony`
> 
> Result:
> 
> ```
> Do you want to update config to the latest version: [y/N]
> N
> Staking mode; node key feea6d16f9073c4a7c8c55cd18822cece32c471bb389b472b884d5cec9e207f5086dc44735d5dc45c4069879383fcf14; -> shard 0
> DNSSyncing: true
> Syncing Port: 3000
> Syncing Zone: t.hmny.io
> GRPC Sync Server: true
> GRPC Sync Server Port:6000 
> GRPC Sync Client: true
> Started RPC server at: 127.0.0.1:9500
> Started WS server at: 127.0.0.1:9800
> ```
> 
> Problem Found: When no config set, migratedFrom was also not set so all version was considered old and was asking for it to be updated.
> Solved: When using default config and not loading from file, set isUsingDefault boolean variable and only ask if false
> 
> ## Test case 2 - Passing dns.port, dns.zone
> Running:
> 
> ```
> ./harmony --dns.port=9001 --dns.zone=127.0.0.1 
> ```
> 
> Result:
> 
> ```
> Staking mode; node key feea6d16f9073c4a7c8c55cd18822cece32c471bb389b472b884d5cec9e207f5086dc44735d5dc45c4069879383fcf14; -> shard 0
> DNSSyncing: true
> Syncing Port: 6001
> Syncing Zone: 127.0.0.1
> GRPC Sync Server: true
> GRPC Sync Server Port:6000 
> GRPC Sync Client: true
> Started RPC server at: 127.0.0.1:9500
> Started WS server at: 127.0.0.1:9800
> ```
> 
> Problems Found: None
> 
> ## Test case 3 Running with config 1.0.2
> Running:
> 
> `./harmony -c config_1_0_2.toml`
> 
> Result:
> 
> ```
> Old config version detected 1.0.2
> Do you want to update config to the latest version: [y/N]
> 
> Staking mode; node key feea6d16f9073c4a7c8c55cd18822cece32c471bb389b472b884d5cec9e207f5086dc44735d5dc45c4069879383fcf14; -> shard 0
> DNSSyncing: true
> Syncing Port: 3000
> Syncing Zone: t.hmny.io
> GRPC Sync Server: true
> GRPC Sync Server Port:6000 
> GRPC Sync Client: true
> Started RPC server at: 127.0.0.1:9500
> Started WS server at: 127.0.0.1:9800
> ```
> 
> Possible Problem: It seems that port DNSSync is actually port - 6000. Not sure if that is intended but we subtract 3000 in the migration, but looking at the above tests, there seem to be somewhere else in the code were there is also - 3000 subtraction
> 
> ## Test case 4 Running with config 1.0.3
> Running:
> 
> `./harmony -c config_1_0_3.toml`
> 
> Result:
> 
> ```
> Old config version detected 1.0.3
> Do you want to update config to the latest version: [y/N]
> 
> Staking mode; node key feea6d16f9073c4a7c8c55cd18822cece32c471bb389b472b884d5cec9e207f5086dc44735d5dc45c4069879383fcf14; -> shard 0
> DNSSyncing: true
> Syncing Port: 3000
> Syncing Zone: t.hmny.io
> GRPC Sync Server: true
> GRPC Sync Server Port:6000 
> GRPC Sync Client: true
> Started RPC server at: 127.0.0.1:9500
> Started WS server at: 127.0.0.1:9800
> ```
> 
> Problems Found: None (Same applies for test 3 and port-6000)
> 
> ## Test case 5 Running with config 1.0.4
> Running:
> 
> `./harmony -c config_1_0_4.toml`
> 
> Result:
> 
> ```
> ld config version detected 1.0.4
> Do you want to update config to the latest version: [y/N]
> 
> Staking mode; node key feea6d16f9073c4a7c8c55cd18822cece32c471bb389b472b884d5cec9e207f5086dc44735d5dc45c4069879383fcf14; -> shard 0
> DNSSyncing: true
> Syncing Port: 3000
> Syncing Zone: t.hmny.io
> GRPC Sync Server: true
> GRPC Sync Server Port:6000 
> GRPC Sync Client: true
> Started RPC server at: 127.0.0.1:9500
> Started WS server at: 127.0.0.1:9800
> ```
> 
> Problems Found: None (Same applies for test 3 and port-6000)

Is the 3000 port being solved? Would you please update the test case result to the correct result?

ntsanov commented:
> Is the 3000 port being solved? Would you please update the test case result to the correct result?

I wasn't sure if that was desired behavior. That's why I marked it as possible problem. I guess It's on me for not formulating the question clearer. To be honest the whole port arthritics  are quite vague to me but it seems that somewhere between parsing the config ( where port is properly -3000) and NewDNSSyncingPeerProvider one more time -3000 is applied. From your comment I guess it is not desired so I removed the logic for subtracting 3000 from from the 1.0.4 migration func and here are the results.

## Test case 3 Running with config 1.0.2
Running:

```./harmony -c config_1_0_2.toml```

Result:
```
Old config version detected 1.0.2
Do you want to update config to the latest version: [y/N]

Staking mode; node key feea6d16f9073c4a7c8c55cd18822cece32c471bb389b472b884d5cec9e207f5086dc44735d5dc45c4069879383fcf14; -> shard 0
DNSSyncing: true
Syncing Port: 6000
Syncing Zone: t.hmny.io
GRPC Sync Server: true
GRPC Sync Server Port:6000
GRPC Sync Client: true
Started RPC server at: 127.0.0.1:9500
Started WS server at: 127.0.0.1:9800
```
Fixed: Syncing port == 6000

## Test case 4 Running with config 1.0.3

Running:

```./harmony -c config_1_0_3.toml```

Result:
```
Old config version detected 1.0.3
Do you want to update config to the latest version: [y/N]

Staking mode; node key feea6d16f9073c4a7c8c55cd18822cece32c471bb389b472b884d5cec9e207f5086dc44735d5dc45c4069879383fcf14; -> shard 0
DNSSyncing: true
Syncing Port: 6000
Syncing Zone: t.hmny.io
GRPC Sync Server: true
GRPC Sync Server Port:6000
GRPC Sync Client: true
Started RPC server at: 127.0.0.1:9500
Started WS server at: 127.0.0.1:9800
```
Fixed: Syncing port == 6000

## Test case 5 Running with config 1.0.4
Running:

```./harmony -c config_1_0_4.toml```

Result:
```
Old config version detected 1.0.4
Do you want to update config to the latest version: [y/N]

Staking mode; node key feea6d16f9073c4a7c8c55cd18822cece32c471bb389b472b884d5cec9e207f5086dc44735d5dc45c4069879383fcf14; -> shard 0
DNSSyncing: true
Syncing Port: 6000
Syncing Zone: t.hmny.io
GRPC Sync Server: true
GRPC Sync Server Port:6000
GRPC Sync Client: true
Started RPC server at: 127.0.0.1:9500
Started WS server at: 127.0.0.1:9800

```
Fixed: Syncing port == 6000

## PS
Is this the desired behavior or I misunderstood something ?

JackyWYX commented on cmd/harmony/main.go:
```diff
@@ -662,7 +661,7 @@ func setupConsensusAndNode(hc harmonyConfig, nodeConfig *nodeconfig.ConfigType)
 		currentNode.SyncingPeerProvider = node.NewLegacySyncingPeerProvider(currentNode)
 	} else {
-		currentNode.SyncingPeerProvider = node.NewDNSSyncingPeerProvider(hc.DNSSync.Zone, legacysync.GetSyncingPort(strconv.Itoa(hc.DNSSync.Port)))
+		currentNode.SyncingPeerProvider = node.NewDNSSyncingPeerProvider(hc.DNSSync.Zone, strconv.Itoa(hc.DNSSync.Port))
```
This line has already been fixed in v1.0.5 PR. I will do another pass to make sure every logic is expected.

JackyWYX reviewed (approved):
Approved! Nice work!! @LeoHChen @rlan35 Take a look?

ntsanov commented:
> Approved! Nice work!! @LeoHChen @rlan35 Take a look?

Wow, thank you, you've been really helpful along the way

JackyWYX commented:
> > Is this PR fully tested by running nodes?
> 
> It's not. It's OK running locally on my end. Let me ask @sophoah for some runtime tests.

@sophoah Can we run node one or two with legacy config file to test this PR out? It shall be all backward compatible.

sophoah commented:
hey guys @JackyWYX @ntsanov, I am testing that PR. Overall things looks ok, I tested 2.0.0 on the below type of nodes : 
- explorer s2 testnet node
- shard 3 external testnet node
- shard 1 internal testnet node

For folks that are already using 1.0.5, that might be an issue, I had some problem running the streamsync code with `unrecognized config version - 1.0.5` For them it's better to redo and use the new 2.0.0 config file directly

see here when the node starts : 
```
Started RPC server at: 127.0.0.1:9500
Started WS server at: 127.0.0.1:9800
Got terminated signal. Gracefully shutting down...
Loaded config version 2.0.0 which is not latest (1.0.5).
Update saved config with `./harmony dumpconfig [config_file]`
Staking mode; node key bafd2e3fa573cf2b6534768aee322a06310b49092a0bc57bab429b692bf745413cef77c05468a54b905026e2c7154893; -> shard 3
Started RPC server at: 127.0.0.1:9500
Started WS server at: 127.0.0.1:9800
```
Is `Loaded config version 2.0.0 which is not latest (1.0.5).`  an expected message ? I had 1.0.5 before i converted it to 2.0.0

`Update saved config with ./harmony dumpconfig [config_file]` needs to be updated to /harmony config dump [config_file]`

by the way what's the difference between Port and ServerPort in the DNSSync section, config dump give me a config of both at 6000, shouldn't the serverPort be at 9000 ? (today 1.0.4 has `DNSPort = 9000`)

JackyWYX commented:
@sophoah Thanks for your very detailed tests.

> For folks that are already using 1.0.5, that might be an issue.

I see. Actually v1.0.5 was not officially announced at the time of implementation so that we decided not to support at the first place. Instead, we squashed the change v1.0.5 along with v2.0.0. (v1.0.5 contains quite some "dirty" code that I am not happy with, which was the initial intention to introduce this bounty) As a developer, I would prefer not to support v1.0.5. But on the ops side, how much effort is expected to get rid of v1.0.5? Would you please help me analysis the impact so that we can probably come to a point of concession?

> Update saved config with ./harmony dumpconfig [config_file] needs to be updated to /harmony config dump [config_file]`.

Yes, need some update to the devops runbook to dump config with the new command. @ntsanov I think it would be also good to support the legacy command to make it backward compatible. What do you think?

> by the way what's the difference between Port and ServerPort in the DNSSync section, config dump give me a config of both at 6000, shouldn't the serverPort be at 9000 ? (today 1.0.4 has DNSPort = 9000)

`Port` is the port for sync at remote DNS server; and `ServerPort` is the local port used as sync server. (@ntsanov Could you please specify the flags a little bit?) The sync port was 9000 at v1.0.4, but actually it is used as 9000-3000=6000. So 6000 for both ports are good as far as I see.

JackyWYX commented:
@ntsanov Some comments for improvement from soph's feedback :point_up_2: Would you please take a look? :)

ntsanov commented:
> @ntsanov Some comments for improvement from soph's feedback  Would you please take a look? :)
Yes of course, thank you @sophoah for the feedback

> Is Loaded config version 2.0.0 which is not latest (1.0.5). an expected message ? I had 1.0.5 before i converted it to 2.0.0

Definitely not. I remember this message but it has been replaced some commits ago. Message in the latest commit is:
```
Old config version detected 1.0.5
Do you want to update config to the latest version: [y/N]
```
I can't explain how you get tihs, but could this be an older checkout ?

>Yes, need some update to the devops runbook to dump config with the new command. @ntsanov I think it would be also good > to support the legacy command to make it backward compatible. What do you think?

I've added the old dumpconfig back in addition to the new "config dump" so backward compatibility should not be an issue any more

> Port is the port for sync at remote DNS server; and ServerPort is the local port used as sync server. (@ntsanov Could you ?
> please specify the flags a little bit?) The sync port was 9000 at v1.0.4, but actually it is used as 9000-3000=6000. So 6000 for > both ports are good as far as I see.

Actually at the moment there is flag only for DNSSync.Port and DNSync.ServerPort gets default or whatever is in the conifg. I  added a flag 
```dns.server-port    "dns sync local server port"```
and change the description of dns.port to "dns sync remote server port". Does that sound ok ?

>I see. Actually v1.0.5 was not officially announced at the time of implementation so that we decided not to support at the first place. Instead, we squashed the change v1.0.5 along with v2.0.0. (v1.0.5 contains quite some "dirty" code that I am not happy with, which was the initial intention to introduce this bounty) As a developer, I would prefer not to support v1.0.5. But on the ops side, how much effort is expected to get rid of v1.0.5? Would you please help me analysis the impact so that we can probably come to a point of concession?

If 1.0.5 will be supported it seems logical to add separate migration for it as well. Please inform if that would be the case

sophoah commented:
Thanks @ntsanov I discussed with @JackyWYX no need to support 1.0.5.

ntsanov commented:
> LegacySyncing

ntsanov commented:
> Also, we need LegacySyncing as a hidden flag

Can you elaborate on that - which flag exactly are we talking about?  No flags were removed, only depricated and a few new added (dns.port, dns.zone, dns.client, dns.server, dns.server-port)

LeoHChen commented on cmd/harmony/config.go:
```diff
@@ -36,18 +36,21 @@ type harmonyConfig struct {
+	Zone          string // replaces: Network.DNSZone
+	LegacySyncing bool   // replaces: Network.LegacySyncing
+	Client        bool   // replaces: Sync.LegacyClient
+	Server        bool   // replaces Sync.LegacyServer
```
minor, missing :

LeoHChen reviewed (commented):
LGTM in general

LeoHChen commented:
have you tested the configuration in systemd service configuration?

sophoah commented:
>have you tested the configuration in systemd service configuration, which is used by most of our validators.
I believe systemd configuration doesn't really matter the most. Validator who are adding custom flag in sytemd is only for temporary settings and we can't possibly know which one they are using. On the config, I believe most are using 1.0.4 or 1.0.3 config https://docs.harmony.one/home/network/validators/node-setup/installing-updating/installing-node/using-binary#option-1-setup-using-config-file-recommended which contains everything that p-ops has tested.

nevertheless, I also just tried 1.0.4 config with the a binary containing this PR and it works just fine.

JackyWYX commented:
@LeoHChen Can we merge?

rlan35 commented:
> @LeoHChen Can we merge?

Please write a instruction about the new config and how to migrate.
